p2p/libp2p/discovery/kadDhtDiscoverer.go

```diff
@@ -230,32 +242,71 @@ func (kdd *KadDhtDiscoverer) IsInterfaceNil() bool {
 	return false
 }
 
-// StartWatchdog start the watchdog, and return the kick channel
-func (kdd *KadDhtDiscoverer) StartWatchdog(timeout time.Duration) chan struct{} {
+// StartWatchdog start the watchdog
+func (kdd *KadDhtDiscoverer) StartWatchdog(timeout time.Duration) error {
 	kdd.mutKadDht.Lock()
 	defer kdd.mutKadDht.Unlock()
 
 	if kdd.contextProvider == nil {
-		return nil
+		return p2p.ErrNilContextProvider
 	}
 
 	if kdd.watchdogKick != nil {
-		return kdd.watchdogKick
+		return p2p.ErrWatchdogAlreadyStarted
+	}
+
+	if timeout < minWatchdogTimeout {
+		return p2p.ErrInvalidDurationProvided
 	}
 
 	kdd.watchdogKick = make(chan struct{})
 	ctx := kdd.contextProvider.Context()
-	go func() {
+	wdCtx, wdCancel := context.WithCancel(ctx)
+	go func(kick <-chan struct{}) {
 		for {
 			select {
 			case <-time.After(timeout):
 				kdd.Resume()
-			case <-ctx.Done():
+			case <-wdCtx.Done():
 				return
-			case <-kdd.watchdogKick:
+			case <-kick:
 			}
 		}
-	}()
+	}(kdd.watchdogKick)
+
+	kdd.watchdogCancel = wdCancel
+	return nil
+}
+
+// StopWatchdog stops the discovery watchdog
+func (kdd *KadDhtDiscoverer) StopWatchdog() error {
+	kdd.mutKadDht.Lock()
+	defer kdd.mutKadDht.Unlock()
 
-	return kdd.watchdogKick
+	if kdd.watchdogCancel == nil {
+		return p2p.ErrWatchdogNotStarted
+	}
+
+	kdd.watchdogCancel()
+	kdd.watchdogCancel = nil
+
+	close(kdd.watchdogKick)
```

Comment:
warning: this might trigger instruction on line 272

Reply:
unlikely, but even if so, the nest select will hit 270 and return